library/ziti.c

```diff
@@ -160,6 +184,222 @@ int load_tls(nf_config *cfg, tls_context **ctx) {
+    }
+
+    mbedtls_ssl_conf_authmode( &conf, MBEDTLS_SSL_VERIFY_NONE );
+    mbedtls_ssl_conf_ca_chain( &conf, &cacert, NULL );
+    mbedtls_ssl_conf_rng( &conf, mbedtls_ctr_drbg_random, &ctr_drbg );
+    // mbedtls_ssl_conf_dbg( &conf, my_debug, stdout );
+
+    if( ( ret = mbedtls_ssl_setup( &ssl, &conf ) ) != 0 ) {
+        ZITI_LOG(ERROR, "mbedtls_ssl_setup returned %d", ret);
+        return ZITI_JWT_VERIFICATION_FAILED;
+    }
+
+    if( ( ret = mbedtls_ssl_set_hostname( &ssl, host ) ) != 0 ) {
+        ZITI_LOG(ERROR, "mbedtls_ssl_set_hostname returned %d", ret);
+        return ZITI_JWT_VERIFICATION_FAILED;
+    }
+
+    mbedtls_ssl_set_bio( &ssl, &server_fd, mbedtls_net_send, mbedtls_net_recv, NULL );
+
+    // Handshake
+    ZITI_LOG(DEBUG, "Performing the SSL/TLS handshake...");
+    while( ( ret = mbedtls_ssl_handshake( &ssl ) ) != 0 ) {
+        if( ret != MBEDTLS_ERR_SSL_WANT_READ && ret != MBEDTLS_ERR_SSL_WANT_WRITE ) {
+            ZITI_LOG(ERROR, "mbedtls_ssl_handshake returned -0x%x\n\n", -ret);
+            return ZITI_JWT_VERIFICATION_FAILED;
+        }
+    }
+
+    // Verify the server certificate
+    ZITI_LOG(DEBUG, "Verifying peer X.509 certificate...");
+    if( ( flags = mbedtls_ssl_get_verify_result( &ssl ) ) != 0 ) {
+        char vrfy_buf[512];
+        ZITI_LOG(ERROR, "X.509 certificate failed verification");
+        mbedtls_x509_crt_verify_info( vrfy_buf, sizeof( vrfy_buf ), "  ! ", flags );
+        ZITI_LOG(ERROR, "%s", vrfy_buf);
+        return ZITI_JWT_VERIFICATION_FAILED;
+    }
+    else {
+        ZITI_LOG(DEBUG, "X.509 certificate is OK");
+    }
+
+    const mbedtls_x509_crt *peerCert = mbedtls_ssl_get_peer_cert( &ssl );
+
+    size_t olen;
+    mbedtls_base64_encode( NULL, 0, &olen, peerCert->raw.p, peerCert->raw.len );  // determine size of buffer we need to allocate
+    char *peerCertString = calloc(1, olen + 1);
+    mbedtls_base64_encode( peerCertString, olen, &olen, peerCert->raw.p, peerCert->raw.len );
+    ZITI_LOG(DEBUG, "peer X.509 certificate is: %s", peerCertString);
+
+    unsigned char output_buf[16000];
+    mbedtls_pk_write_pubkey_pem( &peerCert->pk, output_buf, 16000 );
+    ZITI_LOG(DEBUG, "peer X.509 public key is: %s", output_buf);
+
+    if (strcmp(ecfg->zejh->alg, "RS256") == 0) {
+
+        unsigned char hash[ZITI_MD_MAX_SIZE];
+
+        const mbedtls_md_info_t *md_info = mbedtls_md_info_from_type(MBEDTLS_MD_SHA256);
+
+        mbedtls_md(md_info, ecfg->jwt_signing_input, strlen(ecfg->jwt_signing_input), hash);
+
+        ZITI_LOG(DEBUG, "ecfg->jwt_sig_len is: %d", ecfg->jwt_sig_len);
+
+        if( ( ret = mbedtls_pk_verify( &peerCert->pk, MBEDTLS_MD_SHA256, hash, 0, ecfg->jwt_sig, ecfg->jwt_sig_len ) ) != 0 ) {
+            ZITI_LOG(ERROR, "mbedtls_pk_verify returned -0x%x\n\n", -ret);
+            return ZITI_JWT_VERIFICATION_FAILED;
+        }
+    }
+    else /* default: */
+    {
+        ZITI_LOG(ERROR, "JWT signing algorithm '%s' is not supported", ecfg->zejh->alg);
```

Comment:
JWT sig algo can be things other than RS256.

I might not have been clear when I was describing the algorithms that need support:

RSA keys always use RS256.

EC sig support is currently: ES256 ES384 ES512

Reply:
added